### ISSUE TITLE
feedback

### ISSUE THREAD

**Issue by kmike** [amueller/futurepast#5]
We're looking (rather lazily) for a library to unify and remove deprecation code in Scrapy, keeping an option of creating it ourselves in mind. This one is nice but have some pieces missing:
1. a warning should be issued when user inherits from a renamed class, not only when an instance of renamed class is created. This is important for frameworks which create instances themselves because it allows to issue warning for user's code, not for the code inside a framework. It also allows to show a warning earlier. Subclasses of subclasses of moved classes shouldn't issue a warning because to fix an error user should update a parent class.
2. warning type should be configurable;
3. `isinstance(obj_of_new_class, OldClass)` should return True in case of renames, to avoid breaking code which relies on these type checks.
4. old class path should be customizable, in case when documented/canonical import name is not the same as a path to implementation (e.g. class is exposed in `__init__.py` to an upper level).

We have all these features implemented in https://github.com/scrapy/scrapy/blob/bd4f156d165f67b44fbdec100cb8442a682aea48/scrapy/utils/deprecate.py#L15, tests: https://github.com/scrapy/scrapy/blob/bd4f156d165f67b44fbdec100cb8442a682aea48/tests/test_utils_deprecate.py.

Are features like that too specific for this library?

**Comment by amueller** [comment]
tldr; yes please, all of this is in scope.

Thanks for the input!
I literally spend about half a day on this so far, and I need to spend more on it to make it actually valuable. I hope to do that soon.
1. Makes sense. Though it's not entirely clear to me how to do that. Metaclasses?
2. You mean setting a global configuration flag to say whether to warn, raise or ignore deprecations? That's on my feature wish-list as well. Or would you like more fine-grained control? Or do you want to pass a class specific for your package? [Which could potentially also be a global configuration?]
3. makes sense. That actually just came up in scikit-learn and I didn't realize what the solution was. Sweet!
4. I'm not sure I understand the solution in scrapy.

**Comment by amueller** [comment]
Why do you only warn on direct subclasses for 1? If the intermediate class is never instantiated itself, the user will never see the deprecation.

**Comment by kmike** [comment]
> Makes sense. Though it's not entirely clear to me how to do that. Metaclasses?

There is so much magic in our solution that I don't understand it anymore, but the implementation is in the link :)

> Or would you like more fine-grained control?

Yeah, I'dlike to have more fine-grained control. There can be several libraries using deprecation library and they may want different defaults. 

> Why do you only warn on direct subclasses for 1? If the intermediate class is never instantiated itself, the user will never see the deprecation.

There is still deprecation warning for the direct subclass, and if user fixes this warning there is nothing to fix in subclasses of this subclass.

**Comment by amueller** [comment]
I have to understand this a bit more but I want to solve it. I would like to understand your magic solution before porting it, too ;) I think having all of it in a common place makes sense. Let me release scikit-learn 0.18.1 first, though ;)
